Realm/RLMApp.mm

```diff
@@ -0,0 +1,178 @@
+            });
+        }];
+    }
+};
+
+id <RLMNetworkTransporting> CocoaNetworkTransport::transport = [RLMNetworkTransport new];
+
+@implementation RLMApp : NSObject
+
++(NSMutableDictionary<NSString *, RLMApp*>*)apps {
+    static NSMutableDictionary<NSString *, RLMApp*>* _apps = nil;
+    if (!_apps) {
+        _apps = [NSMutableDictionary new];
+    }
+    return _apps;
+}
+
+-(instancetype) initWithAppId:(NSString *) appId configuration:(RLMAppConfiguration *)configuration {
+    if (self = [super init]) {
+        app::App::Config boundConfiguration = {
+            .app_id = [appId cStringUsingEncoding:NSUTF8StringEncoding]
+        };
+        boundConfiguration.transport_generator = []{
+            return std::unique_ptr<app::GenericNetworkTransport>(new CocoaNetworkTransport);
+        };
+        if (configuration) {
+            if (configuration.baseURL) {
+                boundConfiguration.base_url = util::Optional<std::string>([configuration.baseURL cStringUsingEncoding:NSUTF8StringEncoding]);
+            }
+            if (configuration.transport) {
+                CocoaNetworkTransport::transport = configuration.transport;
+            }
+            boundConfiguration.local_app_name = std::string([configuration.localAppName cStringUsingEncoding:NSUTF8StringEncoding]);
+            boundConfiguration.local_app_version = std::string([configuration.localAppVersion cStringUsingEncoding:NSUTF8StringEncoding]);
+            boundConfiguration.default_request_timeout_ms = (uint64_t)configuration.defaultRequestTimeoutMS;
+        }
+        self->_app = std::make_shared<app::App>(boundConfiguration);
+        [[RLMApp apps] setValue:self forKey:appId];
+        return self;
+    }
+    return nil;
+}
+
++(instancetype) app:(NSString *) appId configuration:(RLMAppConfiguration *)configuration {
+    return [[RLMApp alloc] initWithAppId:appId configuration:configuration];
+}
+
+- (NSDictionary *)allUsers {
+    NSArray *allUsers = [[RLMSyncManager sharedManager] _allUsers];
+    return [NSDictionary dictionaryWithObjects:allUsers
+                                       forKeys:[allUsers valueForKey:@"identity"]];
+}
+
+- (RLMSyncUser *)currentUser {
+    return [[RLMSyncManager sharedManager] _currentUser];
+}
+
+/**
+ Convert an object store AppError to an NSError.
+ */
+static NSError* RLMAppErrorToNSError(const app::AppError& appError) {
+    return [[NSError alloc] initWithDomain:@(appError.error_code.category().name())
+                                      code:appError.error_code.value()
+                                  userInfo:@{
+        @(appError.error_code.category().name()) : @(appError.error_code.message().data())
+    }];
+}
+
+-(void) loginWithCredential:(RLMAppCredentials *)credentials
+          completionHandler:(RLMUserCompletionBlock)completionHandler {
+    self->_app->login_with_credentials(*credentials.appCredentials, ^(std::shared_ptr<SyncUser> user, util::Optional<app::AppError> error) {
```

Comment:
We don't normally self-qualify ivar access as the leading underscore already indicates it's an ivar.